internal/provider/resource_application.go

```diff
@@ -2,340 +2,536 @@ package provider
-	if _, exist := d.GetOk("charm.0.revision"); !exist {
-		revision = -1
+	// if expose is nil, it was not defined
+	var exposeSlice []nestedExpose
+	resp.Diagnostics.Append(plan.Expose.ElementsAs(ctx, &exposeSlice, false)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	var expose map[string]interface{} = nil
+	if len(exposeSlice) == 1 {
+		expose = exposeSlice[0].transformToMapStringInterface()
 	}
 
-	var parsedConstraints constraints.Value = constraints.Value{}
-	readConstraints := d.Get("constraints").(string)
-	if readConstraints != "" {
-		parsedConstraints, err = constraints.Parse(readConstraints)
+	var parsedConstraints = constraints.Value{}
+	if plan.Constraints.ValueString() != "" {
+		parsedConstraints, err = constraints.Parse(plan.Constraints.ValueString())
 		if err != nil {
-			return diag.FromErr(err)
+			resp.Diagnostics.AddError("Input Error", fmt.Sprintf("Unable to parse constraints, go error: %s", err))
 		}
 	}
-
-	response, err := client.Applications.CreateApplication(&juju.CreateApplicationInput{
-		ApplicationName: name,
-		ModelUUID:       modelUUID,
+	response, err := r.client.Applications.CreateApplication(&juju.CreateApplicationInput{
+		ApplicationName: plan.ApplicationName.ValueString(),
+		ModelUUID:       modelInfo.UUID,
 		CharmName:       charmName,
 		CharmChannel:    channel,
 		CharmRevision:   revision,
 		CharmSeries:     series,
-		Units:           units,
+		Units:           int(plan.UnitCount.ValueInt64()),
 		Config:          configField,
 		Constraints:     parsedConstraints,
-		Trust:           trust,
+		Trust:           plan.Trust.ValueBool(),
 		Expose:          expose,
-		Placement:       placement,
+		Placement:       plan.Placement.ValueString(),
 	})
-
 	if err != nil {
-		return diag.FromErr(err)
-	}
-
-	// These values can be computed, and so set from the response.
-	if err = d.Set("name", response.AppName); err != nil {
-		return diag.FromErr(err)
-	}
-
-	charm["revision"] = response.Revision
-	charm["series"] = response.Series
-	if err = d.Set("charm", []map[string]interface{}{charm}); err != nil {
-		return diag.FromErr(err)
-	}
-
-	id := fmt.Sprintf("%s:%s", modelName, response.AppName)
-	d.SetId(id)
-
-	return resourceApplicationRead(ctx, d, meta)
+		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to create application, got error: %s", err))
+		return
+	}
+	r.trace(fmt.Sprintf("create application resource %q", charmName))
+
+	// Faking this plan until the next Read() is done. The value methods used
+	// will put the actual plan values, or default values for the type, e.g.
+	// false and "". Otherwise, there are terraform errors. If we attempt a
+	// ReadApplication, directly after the CreateApplication, application not
+	// found is likely.
+	plan.Constraints = types.StringValue(plan.Constraints.ValueString())
+	plan.Principal = types.BoolValue(plan.Principal.ValueBool())
+	plan.Placement = types.StringValue(plan.Placement.ValueString())
```

Comment:
I'm not entirely sure what's going on here ☝️  why do we need to set these in the Create? if this is because that these fields are "computed", then why aren't we setting the other computed fields, such as, "trust", "units"? I guess I'm not sure what are the terraform errors the comment is talking about

Reply:
It's because they are listed as computed. "Computed" means they must exist even if optional. With trust and units, we know the answer before the application is created. However we do not know if an application is a principal one until deployed. Nor do we know for sure all locations of an application before its deployed. As per the code comment, there is a high likelihood of the application not being found if we attempt to get the data too quickly after deploy as happened during development. 

Try commenting out line 383 to see the errors. 

Comment in code updated.